### DISCUSSION TITLE
Handling `null` and `undefined`

### DISCUSSION THREAD

**Discussion by richardboehme** [ruby/ruby.wasm#26] [General]
Currently it's not possible to determine if a `JS::Object` wrapping `null` or `undefined` is `null`/`undefined`. The method for checking other types (`JS#is_a?`) does not work on those types.

```ruby
null_value = JS::eval('return null')
p JS.is_a?(null_value, JS.global[:null])
# returns false because `JS.global[:null]` is undefined so this cannot work

undefined_value = JS::eval('return undefined')
p JS.is_a?(undefined_value, JS.global[:undefined])
# returns false because `typeof undefined` is 'undefined' and not 'function'
```

I think there are two options for handling this situation:

1. Convert JS `null` to Ruby `nil` and fix `JS#is_a?` for `undefined`.
2. Fix `JS#is_a?` for `null` and `undefined` and keep both wrapped in a `JS::Object`.

I personally think the advantage of 1. is that you can write code like:

```ruby
if js_value
  # do something
else
  # do something else
end
```

However `undefined` is also used by a lot of JS code and you would still have to write code like this:

```ruby
if JS.is_a?(js_object, JS.global[:undefined])
  # do something
else
  # do something else
end
```

The advantage of the second approach is that both cases are handled in the same way. However `JS.global[:null]` is not defined so we'd have to define what value to pass into `JS#is_a?`.

What are your opinions on this?

**Comment by kateinoigakukun** [comment]
> Option 1. Convert JS null to Ruby nil and fix JS#is_a? for undefined.
> Option 2. Fix JS#is_a? for null and undefined and keep both wrapped in a JS::Object.

`JS.is_a?` method is not designed to test whether a value is equal to another value, and it's just an alias of `instanceof` operator in JS. So I don't want to change its behavior.

As a third option, I've added `==` / `eql?` / `strictly_eql?` methods to `JS::Object` to be able to test a JS object is `null` or `undefined`.  https://github.com/ruby/ruby.wasm/pull/28

It would allow you to write null/undefined check by:

```ruby
JS_NULL = JS.eval("return null")
JS_UNDEFINED = JS.eval("return undefined")

if js_value == JS_NULL
  puts "null"
elsif js_value == JS_UNDEFINED
  puts "undefined"
end
```

Does it make sense to you?

**Reply by richardboehme** [reply]
This sounds great! What do you think about pre-defined constants for null and undefined (Like JS::Null and JS::Undefined)? It may not be too useful but I feel like those comparisons are used quite often so they might be helpful.

**Reply by kateinoigakukun** [reply]
I think it's a valuable improvement! Are you interested in making a PR for adding them?

**Reply by richardboehme** [reply]
Yeah definitely! Gonna look into this on Monday.

**Reply by richardboehme** [reply]
I tried out some stuff but it seems like I cannot get it to work. If I understand correctly, the JS object that gets created on initialization is not valid anymore when we reference the constant from JS. The code I tried to use is:
```c
rb_define_const(rb_mJS, "NULL", _rb_js_eval_js(0, rb_str_new_cstr("return null")));
```
When referencing `JS:NULL` from ruby code I get the following error:
```
Uncaught RuntimeError: memory access out of bounds
```

Maybe you have another idea how we could implement this?

Alternatively we could follow the approach of `JS.global` and define `JS.null` as a module function. We may be able to cache the `VALUE` to avoid reevaluating the JS code. What do you think?

**Reply by kateinoigakukun** [reply]
Thank you for your try! Hmm, it's weird. Could you give me a stack trace where the OOB access happens?

**Reply by richardboehme** [reply]
This is the stack trace I got... The code I used is `puts JS::NULL`. I'm new to Ruby C-Extension stuff so maybe I messed something up there? I tried just setting a string as a constant and this worked though.

![image](https://user-images.githubusercontent.com/61323346/177196761-f7ce1a57-4f0e-4e97-9fb2-994a9cc67219.png)

**Reply by kateinoigakukun** [reply]
Sorry for my slow reply. I've added null and undefined constants by https://github.com/ruby/ruby.wasm/pull/30
Could you try it?

**Reply by richardboehme** [reply]
Hey! You do not have to apologize for slow replies! This is open source and I'm so grateful that you even put your time into this :heart:  

I tried your patch and sadly it does not seem to work for me. Comparing like in your tests works, but when trying to print the constant (`p JS::NULL`) it throws the following error:
```
Uncaught Error: eval:103:in `p': method `inspect' called on hidden T_DATA object (0x02369eb0 flags=0xc) (NotImplementedError)
eval:103:in `<main>'
    at checkStatusTag (browser.umd.js:1449:21)
    at evalRbCode (browser.umd.js:1462:7)
    at RubyVM.eval (browser.umd.js:1258:18)
    at HTMLDocument.<anonymous> (application.rb?t=1657540653298&import:1:69)
    at main (main.js?t=1657540653298:23:14)
```
I'm a bit confused by the error message.... I tried to lookup what it could be and stumbled upon a ruby/openssl issue (see [here](https://github.com/ruby/openssl/issues/464)). I'm not sure if it's really related but they fixed it by marking the VALUE. I tried adding a `rb_gc_mark` to `_rb_js_eval_js_cstr` but this just leads to another error:

```
0a2dd336:0x2836b5 Uncaught RuntimeError: table index is out of bounds
    at gc_mark_ptr (0a2dd336:0x2836b5)
    at rb_gc_mark (0a2dd336:0x28464d)
    at _rb_js_eval_js_cstr (0a2dd336:0x64a6)
    at Init_js (0a2dd336:0x5a88)
    at require_internal (0a2dd336:0x3879a6)
    at rb_f_require (0a2dd336:0x38739f)
    at ractor_safe_call_cfunc_1 (0a2dd336:0x5f4569)
    at vm_call_cfunc_with_frame (0a2dd336:0x5e640d)
    at vm_call_method_each_type (0a2dd336:0x5e4724)
    at vm_call_method (0a2dd336:0x5ae24b)
```

At this point I'm out of ideas. Do you have an idea how we can fix this issue?

**Reply by kateinoigakukun** [reply]
Ah, I see. `puts` was the error cause. Let me investigate further later.
